internal/provider/provider.go

```diff
@@ -0,0 +1,129 @@
+package ably_control
+
+import (
+	"context"
+	"os"
+
+	ably_control_go "github.com/ably/ably-control-go"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+func New() tfsdk.Provider {
+	return &provider{}
+}
+
+type provider struct {
+	configured bool
+	client     *ably_control_go.Client
+}
+
+// GetSchema
+func (p *provider) GetSchema(_ context.Context) (tfsdk.Schema, diag.Diagnostics) {
+	return tfsdk.Schema{
+		Attributes: map[string]tfsdk.Attribute{
+			"token": {
+				Type:     types.StringType,
+				Sensitive: true,
+				Required: true,
+			},
+			"url": {
+				Type:      types.StringType,
+				Required: true,
+			},
+		},
+	}, nil
+}
+
+// Provider schema struct
+type providerData struct {
+	Token types.String `tfsdk:"token"`
+	Url types.String `tfsdk:"url"`
+}
+
+func (p *provider) Configure(ctx context.Context, req tfsdk.ConfigureProviderRequest, resp *tfsdk.ConfigureProviderResponse) {
+	// Retrieve provider data from configuration
+	var config providerData
+	diags := req.Config.Get(ctx, &config)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// User must provide a Ably token to the provider
+	var token string
+	if config.Token.Unknown {
+		// Cannot connect to client with an unknown value
+		resp.Diagnostics.AddWarning(
+			"Unable to create client",
+			"Ably Token required",
+		)
+		return
+	}
+
+	if config.Token.Null {
+		token = os.Getenv("ABLY_ACCOUNT_TOKEN")
+	} else {
+		token = config.Token.Value
+	}
+
+	if token == "" {
+		// Error vs warning - empty value must stop execution
+		resp.Diagnostics.AddError(
+			"Unable to find Ably token",
+			"Username cannot be an empty string",
```

Comment:
I am not sure what "username" is referred to here.

Reply:
Fixed